### PR TITLE
Workaround optimizer bug

### PR DESCRIPTION
**Pull request by stefan-koch-sociomantic**
This commit works around a bug in dmd where it'll
pack two distinct values into one register,
which does not work with the current x87 codegen

### CODE REVIEW

**Comment by tiyash-basu-sociomantic** [comment]
Upon trying it out, this workaround does avert the aforementioned dmd bug. Thanks @stefan-koch-sociomantic!

**Comment by Geod24** [comment]
Is there an issue number for this?

**Comment by stefan-koch-sociomantic** [comment]
@Geod24 not yet.

The dmd bug is introduced in https://github.com/dlang/dmd/pull/6192

**Comment by stefan-koch-sociomantic** [comment]
@Geod24 now there is:
https://issues.dlang.org/show_bug.cgi?id=19343

**Comment by don-clugston-sociomantic** [comment]
I'm not sure about the meaning of the comment in the code.
```d
   // x_ptr is needed to create a seperate alias to x
    // which the optimizer cannot see through
```
Does it need to be in that exact form? Can we get rid of one or more of the `size_t`s?
It'd be much nicer if it was just
`auto  x_ptr = &x;`
but I presume that doesn't actually work.

File: src/ocean/math/IEEE.d [hunk 1]
```diff
@@ -1718,21 +1718,27 @@ real NaN(ulong payload)
 ulong getNaNPayload(real x)
 {
     verify(!!isNaN(x));
+    // x_ptr is needed to create a seperate alias to x
```
Comment (iain-buclaw-sociomantic) [inline src/ocean/math/IEEE.d]: Spelling: separate

File: src/ocean/math/IEEE.d [hunk 2]
```diff
@@ -1718,21 +1718,27 @@ real NaN(ulong payload)
...
+    // which the optimizer cannot see through
+    // this will prevent an optimisation which
```
Comment (iain-buclaw-sociomantic) [inline src/ocean/math/IEEE.d]: You use optimize on one line, then optimise on the next.

**Comment by iain-buclaw-sociomantic** [comment]
Can't you just use union access?

**Comment by stefan-koch-sociomantic** [comment]
@don-clugston-sociomantic actually that does work.
Amended.

**Review by matthias-wende-sociomantic — APPROVED** [review]
:+1:

**Review by don-clugston-sociomantic — APPROVED** [review]
LGTM

**Pull request merged**